### PR TITLE
Page parent selector with ComboboxControl.

### LINKED ISSUE #9441: My Parent Pages Are Still Missing In Gutenberg
My parent categories are still not displaying when I am building a page using Gutenberg. When I switch to the Classic Editor, the categories are all there.

I read on a few closed threads that this particular issue would be fixed with the next Gutenberg update (3.0 I believe), which seemed as if that was suppose to be sometime last week. I have deactivated and reinstalled my plug-ins, including Gutenberg, but I am still having the same issue.

When can I expect to see this issue resolved?

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Description
Use the combox selector to provide a searchable page selector.

Fixes https://github.com/WordPress/gutenberg/issues/9441.

Supersedes https://github.com/WordPress/gutenberg/pull/16666. 

Based on work in https://github.com/WordPress/gutenberg/pull/23237.

## Screenshots 
screenshot

Note: screenshot does not include loading indicator which I added a bit later. the logic for loading state needs more work. I found that if I passed false to downshift in initialSelectedItem then later supplied the correct value, the correct value was not shown.  I will work on this issue soon.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by afercia** [comment]
Tested from an accessibility perspective, here's some considerations that apply also to the related PR for the post author, see #23237 

The main issue is that Downshift doesn't work well with Safari and VoiceOver. As also @jeryj commented on https://github.com/WordPress/gutenberg/pull/19657#issuecomment-637656711, Downshift implements the ARIA 1.1 combobox pattern, which is different from the ARIA 1.0 one. The ARIA 1.1 combobox pattern is buggy and undergoing revision. Not a big surprise it's not supported by Safari + VoiceOver.

When landing on the input field, VoiceOver doesn't announce it's a combobox. It just says it's a text field. Expected: it should announce it's a combo box. Without this information, VoiceOver users won't have a clue how to interact with the UI:

downshift no combobox Safari VoiceOver

Instead, the ARIA 1.0 combobox pattern, which is used for example for the Tags autocompleter, works as expected:

tags autocompleter combobox

NVDA and JAWS on Windows seem to work with both patterns, The 1.1 pattern works well and it's announced as expected:

downshift windows correct announcement

So, the problem occurs only on macOS Safari + VoiceOver (can't test on iOS). Although it's specific to one browser / screen reader combination, I'm not sure we can leave out this combo users.

As mentioned a few times in previous related issues, see for example https://github.com/WordPress/gutenberg/pull/16666#issuecomment-518795243 and following comments, I'm not sure why to use Downshift in the first place. Gutenberg already has an autocompleter (the one used for the Tags) that works well across all browsers / screen readers. It's tested "on the field" since it was released and it's a reliable implementation. It would just need to be abstracted and made reusable for other use cases. 

@adamsilverstein when you have a chance, would you like to shed some light whether there's a specific technical reason to use Downshift?

Given the Downshift-powered implementation doesn't work with one of the major browser/ screen reader combinations _and_ the ARIA 1.1 combobox pattern is known to be not very well supported, I'm very hesitant on using Downshift in the first place.

**Other considerations:**

1
The Tags autocompleter does announce the number of results:

tags autocompleter search results

Instead, the Downshift-powered combobox in this PR does not. Same in #23237: no announcement.

Oddly, even when there are results, sometimes screen readers announce a message "No results are available." This message appears to be used internally by Downshift (which uses its own aria-live region) and it doesn't work correctly. We should prevent this message from being announced as it's totally misleading. Instead, a proper announcement of the search results should be added, like the Tags autocompleter.

2
Accuracy of the search results.

While testing, I noticed the search results aren't always accurate, sometimes even confusing, while typing in the input field. For example, one of my pages title is "Page Image Alignment". While typing "page i", the page in question is shown in teh search results:

downshift search results accuracy

But when I keep typing "page imag" in the input field, the related page completely disappears from the search results: no results at all:

downshift search results accuracy 02

I'm not sure what triggers this behavior but I'd think it's potentially very confusing for users.

**Comment by adamsilverstein** [comment]
@afercia Thank you for the feedback as always!

> @adamsilverstein when you have a chance, would you like to shed some light whether there's a specific technical reason to use Downshift?

Primarily because it is included in Gutenberg already in https://github.com/WordPress/gutenberg/pull/17926, plus the upstream developer seemed actively willing to address issues we raised. 

I have also tried implementations using `accessible-autocomplete` which worked really well, but my PR languished for many, many months without any core developer feedback. Then Downshift was added on the other ticket (despite your earlier testing showing it's pitfalls/limitations) so **I understood that to mean the core team had settled on using Downshift**. Recently I noticed the font selector was changed and no longer uses Downshift ❓ 🤷 

> Gutenberg already has an autocompleter (the one used for the Tags) that works well across all browsers / screen readers.

It is good to know our autocompleter works well. I will review the possibility of using the tag auto-completer again. As I recall from my earlier pass, there are some interactions that don't work well with that component used in a select context, however it is worth trying again.

> While testing, I noticed the search results aren't always accurate, sometimes even confusing, while typing in the input field. For example, one of my pages title is "Page Image Alignment". While typing "page i", the page in question is shown in teh search results:

This is likely a code bug, the interactions here are a bit complex and I can see how a race condition bug could be cropping up. I will work to address this once we settle on the approach.

> When landing on the input field, VoiceOver doesn't announce it's a combobox.

This sounds like a Downshift bug. Might be worth opening an upstream issue if we continue using the library.

**Comment by adamsilverstein** [comment]
cc: @epiqueras Is Gutenberg still using Downshift? I see the font selector usage was pulled, do you have any background on that decision?

**Comment by afercia** [comment]
Thanks @adamsilverstein ! To my understanding, the font selector still uses Downshift but it's configured to work as a "simple dropdown" with no auto-completion.

The real problem here is that when Downshift is used as an ARIA combobox (input field with auto-completer and suggestions dropdown) it doesn't work at all with Safari and VoiceOver. I wouldn't say this is an upstream "bug". It's a feature, as Downshift opted to use the ARIA 1.1 pattern that is not supported by VoiceOver.

I'm very hesitant to use Downshift for this use case as the decision made by the Downshift dev team made it incompatible with one of the major browser / screen reader combinations.

**Comment by youknowriad** [comment]
I updated this with the new ComboboxControl

**Comment by adamsilverstein** [comment]
I'll give this a test, nice work @youknowriad!

**Comment by youknowriad** [comment]
@adamsilverstein I'll take a look, I think we need this #25692 as well to be able to reset the value here.

**Comment by adamsilverstein** [comment]
@youknowriad Should we work on a loading indicator for this and authors in a separate PR? Should I start with an issue and get design feedback?

**Comment by youknowriad** [comment]
@adamsilverstein Yes, please do. Whatever works for you an issue or a PR. Both can drag designers.

**Comment by adamsilverstein** [comment]
@youknowriad - created follow up for loading indicator -  https://github.com/WordPress/gutenberg/issues/25798

I noticed when we merged the author changes, we now use the `ComboboxControl` in all cases. We might want to consider using the standard HTML select element when the total number of items is relatively small, eg maybe up to 10, 20 or even 50 items. In these cases it is probably a better user experience (and likely better for accessibility) to show a traditional select. For authors in particular, many sites may only have a few users, so a searchable component is overly complex for their needs.

Shall I open a follow up issue to change this for the authors select, and we can discuss the ideal number to use there?

**Comment by afercia** [comment]
> For authors in particular, many sites may only have a few users, so a searchable component is overly complex for their needs.

I'd be totally in favor of this. No need for a (relatively) complex search UI when a site has only, say, 5 users.

Less likely to happen for pages but yes, there may be sites with no more than 10 pages (or 20, or whatever the agreed limit is).

**Comment by youknowriad** [comment]
> For authors in particular, many sites may only have a few users, so a searchable component is overly complex for their needs.

This would require fetching the number of available pages before rendering which I'm not sure how to do now. It doesn't seem like we have a dedicated data action for that. I'd leave it separate for now.

I made some updates here, I'd love some testing before shipping.

**Review by ntsekouras — COMMENTED** [review]
Code wise and when tested this feels and looks good 👍 - I have left a couple small comments though. I haven't approved, only for wanting to know about my second comment `postTypeSlug` related.

File: packages/editor/src/components/page-attributes/parent.js [hunk 1]
```diff
@@ -1,92 +1,148 @@
...
+		? post.title.rendered
+		: `#${ post.id } (${ __( 'no title' ) })`;
+}
+
+export function PageAttributesParent() {
+	const { editPost } = useDispatch( 'core/editor' );
+	const { parentPost, parentPostId } = useSelect( ( select ) => {
+		const { getEntityRecord } = select( 'core' );
+		const { getEditedPostAttribute } = select( 'core/editor' );
+		const postTypeSlug = getEditedPostAttribute( 'type' );
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/editor/src/components/page-attributes/parent.js]: Good catch, we don't really need two useSelect hooks at all.

**Comment by youknowriad** [comment]
In aa701ad I discovered an issue unrelated with this PR but made visible where sometimes getEntityRecords don't return the required values. More details here https://wordpress.slack.com/archives/C02RQC26G/p1602586210264400 

I'm choosing to deoptimize the `_fields` filtering a little bit, I don't see any other solution on the frontend side.

File: packages/editor/src/components/page-attributes/parent.js [hunk 2]
```diff
@@ -1,92 +1,148 @@
...
+				exclude: postId,
+				parent_exclude: postId,
+				orderby: 'menu_order',
+				order: 'asc',
+				_fields: 'id,title,parent',
+			};
+			if ( parentPost && fieldValue && '' !== fieldValue ) {
+				query.search = fieldValue;
+			}
+			const theParentID = getEditedPostAttribute( 'parent' );
```
Comment (ntsekouras) [inline packages/editor/src/components/page-attributes/parent.js:66]: This is still duplicate. `theParentID` and `pageId` are the same and returned under two fields `parentPostId` and `parent` respectively.

**Comment by adamsilverstein** [comment]
@rianrietveld retying to test the latest version against WordPress trunk, I do not see the pages loading/searching when I type, or the initial page added to the list when my parent is outside the original selection. 

I am testing with several hundred pages, several levels deep, ie:

`wp post generate --post_type=page --max_depth=4 --count=500`

Is it working for you?

**Comment by adamsilverstein** [comment]
> I'd be totally in favor of this. No need for a (relatively) complex search UI when a site has only, say, 5 users.

@afercia - see https://github.com/WordPress/gutenberg/issues/26077

**Comment by youknowriad** [comment]
@adamsilverstein It worked great for me. The command above did create scheduled pages though (not published) so I had to force the date.

**Review by ntsekouras — APPROVED** [review]
LGTM 👍 - CI has a failing test (timing problem) that was in other PRs as well. It's not related and can be handled in a separate PR.

**Comment by adamsilverstein** [comment]
> @adamsilverstein It worked great for me. The command above did create scheduled pages though (not published) so I had to force the date.

I'll give it another test.

**Comment by SJNBham** [comment]
We only see our Homepage as an option in the Parent page combo field now and no search results are returned as of Gutenberg 9.2.0. We have about 1,500 pages on our site. We see the full list of our pages using Gutenberg 9.1.1 (although it takes about 10 seconds for the Parent Page dropdown field to appear when creating a new page. Using Edge (Chromium) and WordPress 5.5.1.

2020-10-21 11_26_43-Add New Page ‹ City of Bellingham — WordPress and 6 more pages - Work - Microsof
2020-10-21 11_26_05-Add New Page ‹ City of Bellingham — WordPress and 6 more pages - Work - Microsof

**Comment by adamsilverstein** [comment]
@youknowriad I can confirm this does not work for me either in wp core trunk - I see the same results as my previous test:

* My initial parent page does not load (unless it is in the initial group).
* No searches are performed, I can only select pages from the initial set.

Note that the user selector still works as expected. I'll try to dig in to see what is happening.

**Comment by youknowriad** [comment]
Thanks for testing and for your help  on this @adamsilverstein 
Let's try to fix this during the beta period. Keep on the loop and I'll try to chime in later.

**Comment by adamsilverstein** [comment]
@youknowriad - followed up with fixes in https://github.com/WordPress/gutenberg/pull/26397

**Pull request merged**